python-packages/smithy-core/smithy_core/aio/types.py

```diff
@@ -271,3 +273,153 @@ async def __anext__(self) -> bytes:
         if data:
             return data
         raise StopAsyncIteration
+
+
+class AsyncBytesProvider:
+    """A buffer that allows chunks of bytes to be exchanged asynchronously.
+
+    Bytes are written in chunks to an internal buffer, that is then drained via an async
+    iterator.
+    """
+
+    def __init__(
+        self, intial_data: bytes | None = None, max_buffered_chunks: int = 16
```

Comment:
It's arbitrary. Ultimately there's only one coroutine reading, and N coroutines writing. If N is greater than 1, the buffer will grow. That said, the reader is going to work very quickly since it just dumps the data into a BytesIO immediately. Writers will likely be slower to push data. And I don't expect there to be that many writers to a single stream anyhow.

Depending on how real world usage ends up, we might be able to improve performance by having separate `Condition`s for readers and writers that share a lock. That may result in the side with fewer participants cycling through their queue faster. But I'm not certain about that, and it seems premature. I'd rather wait until we have actual benchmarks on a live service before adding that extra complication.